### PR TITLE
feat(core): allow asset bundling on docker remote host / docker in docker

### LINKED ISSUE #8799: [core] allow asset bundling on docker remote host / docker in docker
Core `AssetStaging` is now exposing a `bundling` option key allowing assets to be built before being uploaded through a docker image.
Unfortunately the docker command is now using volume mount flags to mount input and output folder, making it impossible to execute if docker is set to run on a remote host or in a docker-in-docker env.
### Use Case

I encountered this problem trying to build a couple of lambdas (and a custom resource provider) written in typescript on a Gitlab CI instance with docker-in-docker or executing docker commands via docker-machine.

### Proposed Solution

My proposal is to create two temporary volumes on target docker env, one for inputs and one for outputs.
Then the first volume can be filled running a `busybox` image as helper and calling `docker cp` to fill the volume.
Once the cp has finished the helper is stopped and the build command is invoked.
After that, if exit code is 0, another helper is started and the `docker cp` is used to copy outputs back to the cdk env.

* [x] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by webratz**
Fixes #8799 

This implements an alternative variant on how to put get files into bundling containers. This is more flexible in its use cases for complex Docker setup scenarios but more complex and slower. Therefore it is not enabled as a default, but as an additional option.

For details to the approach please refer to the linked issue.

----

### All Submissions:

* [X] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md)

### Adding new Construct Runtime Dependencies:

* [ ] This PR adds new construct runtime dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md/#adding-construct-runtime-dependencies)

### New Features

* [X] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md)?
	* [x] Did you use `yarn integ` to deploy the infrastructure and generate the snapshot (i.e. `yarn integ` without `--dry-run`)?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by mrgrain** [comment]
This is great! What I'm currently wondering is if this should be a prop, feature flag or ENV variable. 🤔 

Having it as a prop feels odd in the sense that it's very environment dependent. But then we do the same with local bundling already. @rix0rrr do you have a view how these kind of options should be enabled/disabled?

**Comment by webratz** [comment]
did you get to a decision on how enabling of this mode should be implemented?

**Comment by mrgrain** [comment]
Still on it sorry. Had a busy week.

**Comment by mrgrain** [comment]
Okay I've arrived at the conclusion that a prop is the best option here. It's the only way to enable other uses as well, so if a team wants to implement this based on an environment variable they could do something like:
```ts
{
   fileCopyVariant: process.env.BUNDLING_FILE_COPY_VARIANT 
}
```
Similar applies to using a Context value.

Still need to think about naming, reviewing and manually testing this locally.

---

Apologies in advance but this PR will likely take me a couple more weeks. Just so you know in case your planning depends on it.

**Comment by webratz** [comment]
Sure, no problem. Feel also free to edit the branch / fork it again or so, if that helps you to speed up things. The overall important thing is the outcome in the end: A more flexible way that allows to use docker bundling in more situations than today.

File: packages/@aws-cdk/core/lib/bundling.ts [hunk 1]
```diff
@@ -532,6 +559,88 @@ export interface DockerBuildOptions {
   readonly targetStage?: string;
 }
 
+/**
+ * Provides a helper container for copying bundling related files to specific input and output volumes
+ */
+export class DockerImageBundlingCopyHelper {
```
[the remark this answers is not in the document]
Comment (mrgrain) [reply packages/@aws-cdk/core/lib/bundling.ts:565]: I'd also suggest this class takes `options` in the constructor and exposes a run() method.
Then you do the same for `AssetStagingBindMount`.

In `AssetStaging.bundle()` you can replace the code with a switch:
```ts
switch (options.assetStagingType) {
  case BundlingFileCopyVariant.VOLUME_COPY:
    new AssetStagingVolumeCopy(options).run();
    break;
  case BundlingFileCopyVariant.BIND_MOUNT:
  default:
    new AssetStagingBindMount(options).run();
    break;
}

File: packages/@aws-cdk/core/lib/asset-staging.ts [hunk 2]
```diff
@@ -461,16 +448,46 @@ export class AssetStaging extends Construct {
...
-          volumesFrom: options.volumesFrom,
+          volumes,
+          volumesFrom,
         });
+
+        if (options.fileCopyVariant?.valueOf() === BundlingFileCopyVariant.DOCKER_COPY.valueOf()) {
+          helperContainer.copyOutputTo(bundleDir);
+          helperContainer.cleanHelperContainer();
+          helperContainer.cleanVolumes();
+        }
```
Comment (mrgrain) [inline packages/@aws-cdk/core/lib/asset-staging.ts:490]: See other comments
```ts
switch (options.assetStagingType) {
  case BundlingFileCopyVariant.VOLUME_COPY:
    new AssetStagingVolumeCopy(options).run();
    break;
  case BundlingFileCopyVariant.BIND_MOUNT:
  default:
    new AssetStagingBindMount(options).run();
    break;
}
```

**Review by mrgrain — CHANGES_REQUESTED** [review]
I'm going back and fourth on the naming of things. Will need to sleep over it.

You can get started on the other changes, but mostly if you could make the build pass (rosetta documentation error), I can use this branch to do a full build and run some proper tests using Docker in Docker

File: packages/@aws-cdk/aws-lambda-go/README.md [hunk 3]
```diff
@@ -284,3 +284,18 @@ and Go only includes dependencies that are used in the executable. So in this ca
...
+
+By default the input and output of Docker based bundling is handled via bind mounts.
+In situtations where this does not work, like Docker-in-Docker setups or when using a remote Docker socket, you can configure an alternative, but slower, variant that also works in these situations.
+
+ ```ts
+import { BundlingFileCopyVariant } from '@aws-cdk/core';
+new go.GoFunction(this, 'GoFunction', {
+  entry: 'app/cmd/api',
+  bundling: {
+       fileCopyVariant: BundlingFileCopyVariant.DOCKER_COPY,
```
[the remark this answers is not in the document]
Comment (webratz) [reply packages/@aws-cdk/aws-lambda-go/README.md:298]: i'm not entirely sure why it fails though, it seems like the import i added to one did not help either.

File: packages/@aws-cdk/core/lib/bundling.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (webratz) [reply packages/@aws-cdk/core/lib/bundling.ts]: no, i think it should be private. was just not sure on how to achieve it there, as it depends on the dockerExec which is also private and only available there

File: packages/@aws-cdk/aws-lambda-go/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (webratz) [reply packages/@aws-cdk/aws-lambda-go/README.md]: should be fixed now. build passes with this.
the structural changes are not included yet, but should be good enough to test the approach now

**Comment by webratz** [comment]
@mrgrain the current status should reflect the requested changes.
i hope i did not miss anything as I did it in a few batches in between other stuff :)

**Comment by mrgrain** [comment]
Full build reference `7ab5371a-dd69-46a8-8c19-3cf0d68654d9`

**Comment by webratz** [comment]
I have run a full local build of CDK also and run it on a sample project on our internal CICD system that has the docker by docker (with mounted docker socket) and all worked fine with the VOLUME_COPY mode.

File: packages/@aws-cdk/core/lib/private/asset-staging.ts [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+
+/**
+ * Bundles files with bind mount as copy method
+ */
+export class AssetStagingBindMount extends AssetStagingBase {
+  /**
+   * Bundle files with bind mount as copy method
+   */
+  public run() {
+    this.options.image.run({
```
Comment (mrgrain) [inline packages/@aws-cdk/core/lib/private/asset-staging.ts:53]: I know this is currently unused but let's get ahead of ourselves.
```suggestion
    return this.options.image.run({
```

File: packages/@aws-cdk/core/test/bundling.test.ts [hunk 5]
```diff
@@ -598,4 +598,5 @@ describe('bundling', () => {
       'cool', 'command',
     ], { stdio: ['ignore', process.stderr, 'inherit'] })).toEqual(true);
   });
+
```
Comment (mrgrain) [inline packages/@aws-cdk/core/test/bundling.test.ts:601]: ;)
```suggestion
```

File: packages/@aws-cdk/core/lib/private/asset-staging.ts [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+
+  if (proc.status !== 0) {
+    if (proc.stdout || proc.stderr) {
+      throw new Error(`[Status ${proc.status}] stdout: ${proc.stdout?.toString().trim()}\n\n\nstderr: ${proc.stderr?.toString().trim()}`);
+    }
+    throw new Error(`${prog} exited with status ${proc.status}`);
+  }
+
+  return proc;
+}
```
Comment (mrgrain) [inline packages/@aws-cdk/core/lib/private/asset-staging.ts:221]:
```suggestion
}

```

File: packages/@aws-cdk/core/lib/bundling.ts [hunk 7]
```diff
@@ -383,6 +409,7 @@ export class DockerImage extends BundlingDockerImage {
     reset(deprecated);
     return result;
   }
+
```
Comment (mrgrain) [inline packages/@aws-cdk/core/lib/bundling.ts:412]:
```suggestion
```

File: packages/@aws-cdk/core/test/docker-stub-cp.sh [hunk 8]
```diff
@@ -0,0 +1,16 @@
+#!/bin/bash
+set -euo pipefail
+# stub for the `docker` executable. it is used as CDK_DOCKER when executing unit
+# tests in `test.staging.ts` This variant is specific for tests that use the docker copy method for files, instead of bind mounts
+
+echo "$@" >> /tmp/docker-stub.input.concat
+echo "$@" > /tmp/docker-stub.input
```
[the remark this answers is not in the document]
Comment (webratz) [reply packages/@aws-cdk/core/test/docker-stub-cp.sh]: made some more adjustments, hope i catched all occurences, but local test looks good.

File: packages/@aws-cdk/core/lib/bundling.ts [hunk 9]
```diff
@@ -108,6 +109,12 @@ export interface BundlingOptions {
    * @default - no networking options
    */
   readonly network?: string;
+
+  /**
+   * The Which option to use to copy the source files to the docker container and output files back
```
Comment (mrgrain) [inline packages/@aws-cdk/core/lib/bundling.ts:114]:
```suggestion
   * The mechanism used to make source files available to the bundling container and to return the bundling output back to the host.
```

File: packages/@aws-cdk/core/lib/bundling.ts [hunk 10]
```diff
@@ -151,6 +158,25 @@ export interface ILocalBundling {
   tryBundle(outputDir: string, options: BundlingOptions): boolean;
 }
 
+/**
+ * The type of file copy that should be used for bundling
```
Comment (mrgrain) [inline packages/@aws-cdk/core/lib/bundling.ts:162]:
```suggestion
 * The mechanism used to make source files available to the bundling container and to return the bundling output back to the host
```

File: packages/@aws-cdk/core/lib/bundling.ts [hunk 11]
```diff
@@ -151,6 +158,25 @@ export interface ILocalBundling {
...
+export enum BundlingFileCopyVariant {
+  /**
+   * Creates temporary volumes and docker containers
+   * This is slower, but works also in more complex situations with remote or shared docker sockets.
+   */
+  VOLUME_COPY = 'VOLUME_COPY',
+
+  /**
+   * The source and output folders will be mounted as bind mount from the host system
+   * This is faster and simpler, but less portable than the other option.
```
Comment (mrgrain) [inline packages/@aws-cdk/core/lib/bundling.ts:173]:
```suggestion
   * This is faster and simpler, but less portable than `VOLUME_COPY`.
```

File: packages/@aws-cdk/core/lib/bundling.ts [hunk 12]
```diff
@@ -151,6 +158,25 @@ export interface ILocalBundling {
...
+   */
+  BIND_MOUNT = 'BIND_MOUNT',
+
```
Comment (mrgrain) [inline packages/@aws-cdk/core/lib/bundling.ts:176]:
```suggestion
  BIND_MOUNT = 'BIND_MOUNT',
```

File: packages/@aws-cdk/core/lib/bundling.ts [hunk 13]
```diff
@@ -151,6 +158,25 @@ export interface ILocalBundling {
   tryBundle(outputDir: string, options: BundlingOptions): boolean;
 }
 
+/**
+ * The type of file copy that should be used for bundling
+ */
+export enum BundlingFileCopyVariant {
```
Comment (mrgrain) [inline packages/@aws-cdk/core/lib/bundling.ts:164]:
```suggestion
export enum BundlingFileAccess {
```

File: packages/@aws-cdk/core/lib/bundling.ts [hunk 14]
```diff
@@ -108,6 +109,12 @@ export interface BundlingOptions {
    * @default - no networking options
    */
   readonly network?: string;
+
+  /**
+   * The Which option to use to copy the source files to the docker container and output files back
+   * @default - BundlingFileCopyVariant.BIND_MOUNT
+   */
+  readonly assetStagingType?: BundlingFileCopyVariant;
```
Comment (mrgrain) [inline packages/@aws-cdk/core/lib/bundling.ts:117]:
```suggestion
   * @default - BundlingFileAccess.BIND_MOUNT
   */
  readonly bundlingFileAccess?: BundlingFileAccess;
```

File: packages/@aws-cdk/core/lib/private/asset-staging.ts [hunk 15]
```diff
@@ -0,0 +1,221 @@
...
+   * Path where the source files are located
+   */
+  readonly sourcePath: string;
+  /**
+   * Path where the output files should be stored
+   */
+  readonly bundleDir: string;
+}
+
+class AssetStagingBase {
```
Comment (mrgrain) [inline packages/@aws-cdk/core/lib/private/asset-staging.ts:21]:
```suggestion
abstract class AssetStagingBundlingBase {
```

File: packages/@aws-cdk/core/lib/private/asset-staging.ts [hunk 16]
```diff
@@ -0,0 +1,221 @@
+import { spawnSync, SpawnSyncOptions } from 'child_process';
+import * as crypto from 'crypto';
+import * as os from 'os';
+import { AssetStaging } from '../asset-staging';
+import { BundlingOptions } from '../bundling';
+
+/**
+ * Options for Docker based bundling of assets
+ */
+interface AssetStagingCopyOptions extends BundlingOptions {
```
Comment (mrgrain) [inline packages/@aws-cdk/core/lib/private/asset-staging.ts:10]:
```suggestion
interface AssetStagingBundlingOptions extends BundlingOptions {
```

File: packages/@aws-cdk/core/lib/private/asset-staging.ts [hunk 17]
```diff
@@ -0,0 +1,221 @@
...
+          : '1000:1000';
+    }
+    return user;
+  }
+}
+
+/**
+ * Bundles files with bind mount as copy method
+ */
+export class AssetStagingBindMount extends AssetStagingBase {
```
Comment (mrgrain) [inline packages/@aws-cdk/core/lib/private/asset-staging.ts:48]:
```suggestion
export class AssetBundlingBindMount extends AssetBundlingBase {
```

File: packages/@aws-cdk/core/lib/private/asset-staging.ts [hunk 18]
```diff
@@ -0,0 +1,221 @@
...
+        ...(this.options.volumes ?? []),
+      ],
+    });
+  }
+}
+
+/**
+ * Provides a helper container for copying bundling related files to specific input and output volumes
+ */
+export class AssetStagingVolumeCopy extends AssetStagingBase {
```
Comment (mrgrain) [inline packages/@aws-cdk/core/lib/private/asset-staging.ts:80]:
```suggestion
export class AssetBundlingVolumeCopy extends AssetBundlingBase {
```

**Review by mrgrain — CHANGES_REQUESTED** [review]
update naming

File: packages/@aws-cdk/aws-lambda-go/README.md [hunk 19]
```diff
@@ -284,3 +284,18 @@ and Go only includes dependencies that are used in the executable. So in this ca
...
 an update to `auth` or `middleware` would only trigger an update to the `cmd/api` Lambda
 Function.
+
+## Docker based bundling in complex Docker configurations
+
+By default the input and output of Docker based bundling is handled via bind mounts.
+In situtations where this does not work, like Docker-in-Docker setups or when using a remote Docker socket, you can configure an alternative, but slower, variant that also works in these situations.
+
+ ```ts
+import { BundlingFileAccess } from '@aws-cdk/core';
```
Comment (mrgrain) [inline packages/@aws-cdk/aws-lambda-go/README.md:294]: Let's move this import to `packages/@aws-cdk/aws-lambda-go/rosetta/default.ts-fixture`
Same for the other packages.

File: packages/@aws-cdk/aws-lambda-go/lib/bundling.ts [hunk 20]
```diff
@@ -2,6 +2,7 @@ import * as os from 'os';
 import * as path from 'path';
 import { Architecture, AssetCode, Code, Runtime } from '@aws-cdk/aws-lambda';
 import * as cdk from '@aws-cdk/core';
+import { BundlingFileAccess } from '@aws-cdk/core';
```
Comment (mrgrain) [inline packages/@aws-cdk/aws-lambda-go/lib/bundling.ts:5]: Use `cdk. BundlingFileAccess` instead
```suggestion
```

File: packages/@aws-cdk/aws-lambda-go/lib/bundling.ts [hunk 21]
```diff
@@ -60,6 +61,12 @@ export interface BundlingProps extends BundlingOptions {
    * The system architecture of the lambda function
    */
   readonly architecture: Architecture;
+
+  /**
+   * Which option to use to copy the source files to the docker container and output files back
+   * @default - BundlingFileAccess.BIND_MOUNT
+   */
+  readonly bundlingFileAccess?: BundlingFileAccess;
```
Comment (mrgrain) [inline packages/@aws-cdk/aws-lambda-go/lib/bundling.ts:69]:
```suggestion
  readonly bundlingFileAccess?: cdk.BundlingFileAccess;
```

File: packages/@aws-cdk/aws-lambda-nodejs/README.md [hunk 22]
```diff
@@ -337,3 +337,17 @@ new nodejs.NodejsFunction(this, 'my-handler', {
...
 If you chose to customize the hash, you will need to make sure it is updated every time the asset
 changes, or otherwise it is possible that some deployments will not be invalidated.
+
+## Docker based bundling in complex Docker configurations
+
+By default the input and output of Docker based bundling is handled via bind mounts.
+In situtations where this does not work, like Docker-in-Docker setups or when using a remote Docker socket, you can configure an alternative, but slower, variant that also works in these situations.
+
+ ```ts
+import { BundlingFileAccess } from '@aws-cdk/core';
```
Comment (mrgrain) [inline packages/@aws-cdk/aws-lambda-nodejs/README.md:347]: import -> rosetta fixture

File: packages/@aws-cdk/aws-lambda-nodejs/lib/bundling.ts [hunk 23]
```diff
@@ -6,6 +6,7 @@ import { PackageInstallation } from './package-installation';
 import { LockFile, PackageManager } from './package-manager';
 import { BundlingOptions, OutputFormat, SourceMapMode } from './types';
 import { exec, extractDependencies, findUp, getTsconfigCompilerOptions } from './util';
+import { BundlingFileAccess } from '@aws-cdk/core';
```
Comment (mrgrain) [inline packages/@aws-cdk/aws-lambda-nodejs/lib/bundling.ts:9]: Use `cdk. BundlingFileAccess` instead
```suggestion
```

File: packages/@aws-cdk/aws-lambda-nodejs/lib/bundling.ts [hunk 24]
```diff
@@ -43,6 +44,12 @@ export interface BundlingProps extends BundlingOptions {
    */
   readonly preCompilation?: boolean
 
+  /**
+   * Which option to use to copy the source files to the docker container and output files back
+   * @default - BundlingFileAccess.BIND_MOUNT
+   */
+  readonly bundlingFileAccess?: BundlingFileAccess;
```
Comment (mrgrain) [inline packages/@aws-cdk/aws-lambda-nodejs/lib/bundling.ts:51]:
```suggestion
  readonly bundlingFileAccess?: cdk.BundlingFileAccess;
```

File: packages/@aws-cdk/aws-lambda-python/README.md [hunk 25]
```diff
@@ -252,3 +252,21 @@ an array of commands to run. Commands are chained with `&&`.
...
+ const entry = '/path/to/function';
+
+new python.PythonFunction(this, 'function', {
+  entry,
+  runtime: Runtime.PYTHON_3_8,
+  bundling: {
+    bundlingFileAccess: BundlingFileAccess.VOLUME_COPY,
+   },
+});
+```
```
Comment (mrgrain) [inline packages/@aws-cdk/aws-lambda-python/README.md:272]:
```suggestion
import -> rosetta and remove some whitespace
```ts
const entry = '/path/to/function';

new python.PythonFunction(this, 'function', {
  entry,
  runtime: Runtime.PYTHON_3_8,
  bundling: {
    bundlingFileAccess: BundlingFileAccess.VOLUME_COPY,
   },
});
```
```

File: packages/@aws-cdk/core/test/private/asset-staging.test.ts [hunk 26]
```diff
@@ -0,0 +1,115 @@
+import * as child_process from 'child_process';
+import * as sinon from 'sinon';
+import { AssetStaging, DockerImage } from '../../lib';
+import { AssetBundlingBindMount, AssetBundlingVolumeCopy } from '../../lib/private/asset-staging';
+
+describe('bundling', () => {
+  afterEach(() => {
+    sinon.restore();
+
```
Comment (mrgrain) [inline packages/@aws-cdk/core/test/private/asset-staging.test.ts:9]:
```suggestion
```

File: packages/@aws-cdk/core/test/private/asset-staging.test.ts [hunk 27]
```diff
@@ -0,0 +1,115 @@
...
+    // actual docker run with bind mount is called
+    expect(spawnSyncStub.calledWith('docker', sinon.match.array.contains([
+      'run', '--rm',
+      '-v',
+      'alpine',
+    ]), { stdio: ['ignore', process.stderr, 'inherit'] })).toEqual(true);
+
+  });
+
+});
```
Comment (mrgrain) [inline packages/@aws-cdk/core/test/private/asset-staging.test.ts:115]:
```suggestion
  });
});
```

File: packages/@aws-cdk/core/test/staging.test.ts [hunk 28]
```diff
@@ -1,21 +1,26 @@
+// import * as child_process from 'child_process';
```
Comment (mrgrain) [inline packages/@aws-cdk/core/test/staging.test.ts:1]:
```suggestion
```

**Review by mrgrain — APPROVED** [review]

**Pull request merged**